draft-packages/popover/KaizenDraft/Popover/Popover.tsx

```diff
@@ -38,6 +38,14 @@ type Position = "start" | "center" | "end"
 
 type Size = "small" | "large"
 
+type BoxOffset =
+  | number
+  | undefined
+  | {
+      xOffset?: number // to ensure a non-breaking change, xOffset can only be a number
+      yOffset?: string
+    }
```

Comment:
Neat! The only thing I wonder if it is worth the more complicated interface of allowing a string as `yOffset` for the extra flexibility that provides?

It sounded like this component might be getting re-designed - what is the timescale for that change?

Reply:
So I've allowed string as it means that people can set calculations etc to make sure it's set just right. Ideally, x would also be string but that would potentially be a breaking change

Reply:
Yep, so you think the extra flexibility justifies the slightly more complex interface (accepting two different types of values for similar fields)?

Reply:
I do, as I think not allowing this flexibility for the y offset would actually significantly decrease its usefulness.

Reply:
@stephwestgarth Could you not enjoy the same flexibility for `xOffset` without a breaking change by making it a union type of `number | string`?

Reply:
Do you have a suggestion of what that y offset number would represent? Would it be `px` or `%`? I think that is one of the reasons why I also left it to the user to specify

Reply:
I think I'd lean towards a `px` value if it's just a number. For a percentage I think I would want to see a name like `yOffsetAsPercentage` or a decimal value like `0.5`.

Reply:
Sorry, I forgot to tag you @sebpearce! 😅

Reply:
Hm, I wonder if that is starting to over-engineer what is meant to be an unblocking change. Applying the [yagni principle](https://martinfowler.com/bliki/Yagni.html), especially as we want to re-engineer how this works entirely so it's not as dumb of a component, is that adding a feature in that won't be used? I like the idea but get concerned about adding in complexity like that for no reason

Reply:
I think we're seeing this from different angles – I thought from the comment below, you meant that you would like to have `xOffset` as a string, but couldn't because it would create a breaking change:
![image](https://user-images.githubusercontent.com/6406263/97516871-35c2d480-19e8-11eb-8167-4d4b9b831039.png)
I didn't consider that this would be adding a feature.

My main concern with any temporary/unblocking change is: how hard is this going to be to climb out of when we do actually want to do that eventual "proper" change? If you feel that it'll be fairly easy, then I say go for it.

Reply:
Ah, I'm on the same page now. To clarify, x offset is what the old version of boxOffset represented. Y offset is entirely new, and adding in extra complexity for something that _hopefully_ won't exist long is a bit of overkill imo.

So my impression of how we'd resolve this in future is it wouldn't be css driven at all (or css would be for minute changes, in which case it would likely be px like x offset); the ideal solution would actually be driven by js (intelligent target positioning to prevent off-screen popovers etc), so how this works would be changed entirely.

Reply:
Sounds good to me!

Reply:
Heh yeah, this was a suggestion from me. I think the `number | undefined | { xOffset: string, yOffset }` is a good middle ground for now. The tradeoff that @thrgamon mentioned is spot on - a slightly more complicated API now to avoid a breaking change for this small fix. My rationale is that popovers need a little more love in the future and I'd rather set aside a breaking change to rethink a few things about this component (like better OOTB position that @stephwestgarth mentioned - probably using something like popperjs).

What I think may have happened is this component was initially made to support one orientation (hence only needing to support xoffset).

Reply:
Hey! So I would have added this property originally, so I can provide some context.

The `boxOffset` property was designed as a way to offset the main popover rectangle from its arrow. eg.

![image](https://user-images.githubusercontent.com/4495057/97530558-80534980-1a06-11eb-9af8-8c12ba7b15c2.png)

I'm not sure if this is still used anymore in perform-ui, now that we have the new zen headers. So it is possible that we could remove it.

With that said, my understanding with this update is that it offsets the entire popover, not just the rectangle. This is different functionality, therefore if this feature ends up staying in, it should be added as a separate property. Maybe `popoverOffset` or `rootOffset`?

> What I think may have happened is this component was initially made to support one orientation (hence only needing to support xoffset).

Almost yeah. Since `boxOffset` only moves the rectangle and content, there'd never be a reason to move the box away from its pointer arrow. If we were ever to support "left" | "right" popover positioning, `boxOffset` would act on the y axis.

I don't have full context of the problem to solve, but as an alternative, maybe a wrapper around the Popover could work? That way you wouldn't need to update kaizen eg.

```
<div style={{ transform: "translateY(20px)" }}><Popover /></div>
```

Or a negative margin could also work - nested transforms can sometimes cause weird aliasing effects.

Reply:
And apologies for the lack of documentation here.

Reply:
Hey @jeremy8883 , so I can see that we also use it in the getRootStyle function, which implies this is a style of the root. I guess that is why we've kept it here, as it is used in both the root offset and this arrow offset.

Reply:
Yeah sorry for the confusion, the reason why it does an offset there, is because it needs to compensate for the change made to the rectangle position. eg. before the compensation transform:

![image](https://user-images.githubusercontent.com/4495057/97533041-9a435b00-1a0b-11eb-9d46-c8cd2d946af8.png)

after the compensation transform:

![image](https://user-images.githubusercontent.com/4495057/97533087-acbd9480-1a0b-11eb-9042-e8f33e83efee.png)

Reply:
So from what you're saying, I think it is in the right place for now given how it currently works, but we need to rethink how we do this in future entirely (which is already on the todo list!)